### PR TITLE
[Cocoa] MediaCapabilities spatial rendering checks fail on capable HW with built-in sound output

### PR DESCRIPTION
**Pull request by eric-carlson**
#### f8df1646f208282f5b000bf105beebd2e536a798
<pre>
[Cocoa] MediaCapabilities spatial rendering checks fail on capable HW with built-in sound output
<a href="https://bugs.webkit.org/show_bug.cgi?id=265614">https://bugs.webkit.org/show_bug.cgi?id=265614</a>
<a href="https://rdar.apple.com/114214068">rdar://114214068</a>

Reviewed by Jer Noble.

On macOS, use SPI got get spatial audio preferences to see if spatial audio playback is
supported. On iOS, switch from using the shared audio presentation output context to
AVAudioSession, which makes it possible to see if the currently audio route has a port
that supports spatial playback.

* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.h:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/AudioToolboxSPI.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setSupportsSpatialAudioPlayback):
(WebCore::PlatformMediaSessionManager::supportsSpatialAudioPlaybackForConfiguration):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::supportsSpatialAudioPlayback):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.h:
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::MediaSessionManagerCocoa):
(WebCore::MediaSessionManagerCocoa::updateSessionState):
(WebCore::MediaSessionManagerCocoa::supportsSpatialAudioPlaybackForConfiguration):
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
(MediaSessionHelper::activeAudioRouteSupportsSpatialPlaybackDidChange):
(MediaSessionHelper::updateActiveAudioRouteSupportsSpatialPlayback):
(MediaSessionHelper::setActiveAudioRouteSupportsSpatialPlayback):
(-[WebMediaSessionHelper initWithCallback:]):
(-[WebMediaSessionHelper spatialPlaybackCapabilitiesChanged:]):
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
(WebCore::MediaSessionManageriOS::activeAudioRouteSupportsSpatialPlaybackDidChange):
(WebCore::MediaSessionManagerCocoa::supportsSpatialAudioPlaybackForConfiguration):
* Source/WebCore/platform/graphics/cocoa/MediaEngineConfigurationFactoryCocoa.cpp:
(WebCore::computeMediaCapabilitiesInfo):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
(WebKit::RemoteMediaSessionHelperProxy::activeAudioRouteSupportsSpatialPlaybackDidChange):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.h:
* Source/WebKit/Shared/mac/MediaFormatReader/CoreMediaWrapped.cpp:
(WebKit::wrapperVTable):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::activeAudioRouteSupportsSpatialPlaybackDidChange):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.messages.in:

Canonical link: <a href="https://commits.webkit.org/271504@main">https://commits.webkit.org/271504@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/fe29386cffcefd04c1b8f17461a6850923db9eed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30927 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25858 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4414 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26107 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24434 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5079 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25435 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26007 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31479 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5146 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3323 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29240 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6745 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6848 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5602 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5666 "Built successfully") | | |

### CODE REVIEW

**Review by jernoble — APPROVED** [review]

File: Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp [hunk 1]
```diff
@@ -435,6 +435,20 @@ void PlatformMediaSessionManager::setIsPlayingToAutomotiveHeadUnit(bool isPlayin
...
+void PlatformMediaSessionManager::setSupportsSpatialAudioPlayback(bool supportsSpatialAudioPlayback)
+{
+    if (supportsSpatialAudioPlayback == m_supportsSpatialAudioPlayback)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER, supportsSpatialAudioPlayback);
+    m_supportsSpatialAudioPlayback = supportsSpatialAudioPlayback;
+}
+
+const std::optional<bool> PlatformMediaSessionManager::supportsSpatialAudioPlaybackForConfiguration(const MediaConfiguration&)
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp]: const here doesn't make much, as the value will be copied anyway.
PlatformMediaSessionManager is a WebContent only object

File: Source/WebCore/platform/audio/PlatformMediaSessionManager.h [hunk 2]
```diff
@@ -152,6 +153,9 @@ class PlatformMediaSessionManager
     WEBCORE_EXPORT void setIsPlayingToAutomotiveHeadUnit(bool);
     bool isPlayingToAutomotiveHeadUnit() const { return m_isPlayingToAutomotiveHeadUnit; }
 
+    WEBCORE_EXPORT void setSupportsSpatialAudioPlayback(bool);
+    virtual const std::optional<bool> supportsSpatialAudioPlaybackForConfiguration(const MediaConfiguration&);
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/PlatformMediaSessionManager.h]: ditto

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm [hunk 3]
```diff
@@ -59,7 +61,7 @@
 #endif // !PLATFORM(MAC)
 
 MediaSessionManagerCocoa::MediaSessionManagerCocoa()
-    : m_nowPlayingManager(platformStrategies()->mediaStrategy().createNowPlayingManager())
+    : m_nowPlayingManager(hasPlatformStrategies() ? platformStrategies()->mediaStrategy().createNowPlayingManager() : nullptr)
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:64]: that doesn't seem to be related to this change

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm [hunk 4]
```diff
@@ -544,6 +546,56 @@
...
+{
+    ASSERT(configuration.audio);
+    if (!configuration.audio)
+        return { false };
+
+    auto supportsSpatialAudioPlayback = this->supportsSpatialAudioPlayback();
+    if (supportsSpatialAudioPlayback.has_value())
+        return supportsSpatialAudioPlayback;
+
+    auto calculateSpatialAudioSupport = [] (const MediaConfiguration& configuration) -> bool {
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm]: all path return bool, so `-> bool` isn't necessary

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm [hunk 5]
```diff
@@ -544,6 +546,56 @@
...
+            return true;
+
+        bool supported = false;
+        for (uint32_t i = 0; i < spatialAudioPreferences.spatialAudioSourceCount; ++i) {
+            auto& source = spatialAudioPreferences.spatialAudioSources[i];
+            if (source == kSpatialAudioSource_Multichannel && channelCount > 2) {
+                supported = true;
+                break;
+            }
+            if (source == kSpatialAudioSource_MonoOrStereo && channelCount >= 1) {
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm]: `&& channelCount` ?

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm [hunk 6]
```diff
@@ -544,6 +546,56 @@
...
+
+        auto channelCount = configuration.audio->channels.toDouble();
+        if (channelCount <= 0)
+            return true;
+
+        bool supported = false;
+        for (uint32_t i = 0; i < spatialAudioPreferences.spatialAudioSourceCount; ++i) {
+            auto& source = spatialAudioPreferences.spatialAudioSources[i];
+            if (source == kSpatialAudioSource_Multichannel && channelCount > 2) {
+                supported = true;
```
Comment (jyavenard) [inline Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm]: could `return true` early rather than break

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm (hunk 3 above)
Comment (eric-carlson) [reply Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:64]: This will crash without this change as `platformStrategies()` returns NULL in the GPU process.

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm (hunk 5 above)
Comment (eric-carlson) [reply Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm]: `channelCount` is a double.

File: Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp (hunk 1 above)
Comment (eric-carlson) [reply Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp]: True, originally it returned a `std::optional<bool>&` so const made sense at that point. I will change it.

File: Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm (hunk 6 above)
Comment (eric-carlson) [reply Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm]: Good idea
